### PR TITLE
feat(sns-ext): add `propose` subcommand

### PR DESCRIPTION
**Pull request by DanielThurau**

### CODE REVIEW

File: extensions/sns/README.md [hunk 1]
```diff
@@ -116,4 +121,39 @@ Remove NNS Root as a co-controller to a dapp canister controlled by the current
 ```
 dfx sns prepare-canisters remove-nns-root rkp4c-7iaaa-aaaaa-aaaca-cai
 dfx sns prepare-canisters remove-nns-root rkp4c-7iaaa-aaaaa-aaaca-cai 6zikg-xaaaa-aaaaa-aabhq-cai
+```
+
+## dfx sns propose
+
+Use the `dfx sns propose` command to submit a CreateServiceNervousSystem NNS proposal according to the
+local configuration file. The local dfx identity must be able to operate (as a controller or hotkey) 
+a staked NNS Neuron to be able to submit an NNS Proposal.
```
Comment (DanielThurau) [inline extensions/sns/README.md:130]:
```suggestion
a staked NNS Neuron to submit an NNS Proposal. 
```

File: extensions/sns/README.md [hunk 2]
```diff
@@ -116,4 +121,39 @@ Remove NNS Root as a co-controller to a dapp canister controlled by the current
...
+
+### Basic usage
+
+``` bash
+dfx sns propose
+```
+
+### Examples
+
+Submit a proposal using a known NeuronId
```
Comment (DanielThurau) [inline extensions/sns/README.md:140]:
```suggestion
Submit a proposal using a known NeuronId.
```

File: extensions/sns/README.md [hunk 3]
```diff
@@ -116,4 +121,39 @@ Remove NNS Root as a co-controller to a dapp canister controlled by the current
...
+```
+dfx sns propose --neuron-id 42 sns_init.yaml 
+```
+
+Submit a proposal using the memo chosen during NNS Neuron creation. This is used in conjunction
+with the current dfx identity to calculate the ICP Ledger Subaccount that backs the NNS Neuron's 
+stake.
+
+```
+dfx sns propose --neuruon-memo 0 sns_init.yaml
```
Comment (ericswanson-dfinity) [inline extensions/sns/README.md:151]:
```suggestion
dfx sns propose --neuron-memo 0 sns_init.yaml 
```

File: extensions/sns/src/commands/propose.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+/// SNS propose command line arguments.
+#[derive(Debug, Parser)]
+#[command(name("propose"))]
+pub struct SnsProposeOpts {
+    /// The id of the neuron with which to make the proposal. The current dfx
+    /// identity must be able to operate this neuron (as a hotkey for instance).
+    #[arg(
+        long,
+        conflicts_with = "neuron_memo",
+        conflicts_with = "test_neuron_proposer"
```
Comment (ericswanson-dfinity) [inline extensions/sns/src/commands/propose.rs:15]:
```suggestion
        group = "neuron-selection"
```

File: extensions/sns/src/commands/propose.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    neuron_id: Option<u64>,
+
+    /// The memo to use in conjunction with the current dfx identity to
+    /// identify the neuron with which to make the proposal. This calculates
+    /// the subaccount address with which the Neuron was created with. The current dfx
+    /// identity must be able to operate this neuron (as a hotkey for instance).
+    #[arg(
+        long,
+        conflicts_with = "neuron_id",
+        conflicts_with = "test_neuron_proposer"
```
Comment (ericswanson-dfinity) [inline extensions/sns/src/commands/propose.rs:26]:
```suggestion
        group = "neuron-selection"
```

File: extensions/sns/src/commands/propose.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    neuron_memo: Option<u64>,
+
+    /// A test only alternative to `--neuron-id` and `--neuron-memo` that works
+    /// on a local dfx server where an NNS is installed with the test neuron. If used
+    /// with the mainnet version of the ic, `dfx sns propose` will return a failure.
+    #[arg(
+        long,
+        default_value = "false",
+        conflicts_with = "neuron_id",
+        conflicts_with = "neuron_memo"
```
Comment (ericswanson-dfinity) [inline extensions/sns/src/commands/propose.rs:37]:
```suggestion
        group = "neuron-selection"
```

File: extensions/sns/src/commands/propose.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            OsString::from("--neuron-id"),
+            OsString::from(neuron_id.to_string()),
+        ]),
+        (None, Some(neuron_memo), false) => args.extend(vec![
+            OsString::from("--neuron-memo"),
+            OsString::from(neuron_memo.to_string()),
+        ]),
+        (None, None, true) => args.extend(vec![OsString::from("--test-neuron-proposer")]),
+        _ => {
+            anyhow::bail!("one of the arguments '--neuron-id <NEURON_ID>' or '--neuron-memo <NEURON_MEMO>' or --test-neuron-proposer must be used");
```
Comment (ericswanson-dfinity) [inline extensions/sns/src/commands/propose.rs:67]:
```suggestion
            unreachable!("no neuron selected");
```

File: extensions/sns/README.md (hunk 3 above)
Comment (DanielThurau) [reply extensions/sns/README.md:151]: 🤦‍♂️ Fixed

File: extensions/sns/src/commands/propose.rs [hunk 8]
```diff
@@ -0,0 +1,76 @@
+use clap::Parser;
+use dfx_extensions_utils::call_extension_bundled_binary;
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+
+/// SNS propose command line arguments.
+#[derive(Debug, Parser)]
+#[command(name("propose"))]
+pub struct SnsProposeOpts {
```
[the remark this answers is not in the document]
Comment (DanielThurau) [reply extensions/sns/src/commands/propose.rs:9]: This is waaaay better. Thank you for the sugggestion!

**Comment by DanielThurau** [comment]
@ericswanson-dfinity could you take another look now that I've addressed comments and added more tests?

File: extensions/sns/e2e/tests/sns.bash [hunk 9]
```diff
@@ -167,3 +168,84 @@ SNS_CONFIG_FILE_NAME="sns.yml"
...
+                };
+            }
+        };
+        neuron_id_or_subaccount = null;
+    })"
+    assert_success
+
+    # Actually submit the proposal
+    run dfx sns propose --neuron-id "${NEURON_ID}" "valid/${SNS_CONFIG_FILE_V2_NAME}"
+    assert_success
```
[the remark this answers is not in the document]
Comment (DanielThurau) [reply extensions/sns/e2e/tests/sns.bash:250]: The command will actually interpret the response from the the governance canister and exit with status 1 if the proposal is not accepted. I was able to verify this in code, and observing this behavior in the test when I failed to have things submitted correctly.

[the remark this answers is not in the document]
Comment (ericswanson-dfinity) [reply extensions/sns/e2e/tests/sns.bash:250]: Does "dfx sns propose" print anything out from the response that it receives?

What I'm looking for is evidence that the command did more than "exit 0"

[the remark this answers is not in the document]
Comment (DanielThurau) [reply extensions/sns/e2e/tests/sns.bash:250]: Yes it prints the proposal id. I can grep for that.

[the remark this answers is not in the document]
Comment (DanielThurau) [reply extensions/sns/e2e/tests/sns.bash:252]: Done. PTAL

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**